main/src/library/Iterator.flix

```diff
@@ -195,108 +174,84 @@ namespace Iterator {
-        let d = new MutDeque(r1);
+    pub def toMutDeque(rc: Region[r1], iter: Iterator[a, ef, r2]): MutDeque[a, r1] \ { ef, Read(r2), Write(r1) } =
+        let d = new MutDeque(rc);
         forEach(x -> MutDeque.pushBack(x, d), iter);
         d
 
     ///
     /// Returns the contents of `iter` as a set. Consumes the entire iterator.
     ///
-    pub def toSet(iter: Iterator[a, r]): Set[a] \ Read(r) with Order[a] =
+    pub def toSet(iter: Iterator[a, ef, r]): Set[a] \ { ef, Read(r) } with Order[a] =
         foldLeft((acc, a) -> Set.insert(a, acc), Set.empty(), iter)
 
-    ///
-    /// Returns a list of every element in `iter` that satisfies the predicate `f`.
-    ///
-    /// Consumes the entire iterator.
-    ///
-    pub def filter(f: a -> Bool \ ef, iter: Iterator[a, r]): List[a] \ { ef, Read(r) } =
-        let Iterator(done, next) = iter;
-        def loop(k) = {
-            if (done())
-                k(Nil)
-            else
-                let x = next();
-                if (f(x)) loop(ks -> k(x :: ks)) else loop(k)
-        };
-        loop(identity)
-
-    ///
-    /// Returns a list with `f` applied to every element in `iter`.
-    ///
-    /// That is, the result is of the form: `f(x1) :: f(x2) :: ...`.
-    ///
-    /// Consumes the entire iterator.
-    ///
-    pub def map(f: a -> b \ ef, iter: Iterator[a, r]): List[b] \ { ef, Read(r) } =
-        let Iterator(done, next) = iter;
-        def loop(k) = {
-            if (done())
-                k(Nil)
-            else
-                let x = f(next());
-                loop(ks -> k(x :: ks))
-        };
-        loop(identity)
-
     ///
     /// Applies `f` to every element of `iter`.
     ///
     /// Consumes the entire iterator.
     ///
-    pub def forEach(f: a -> Unit \ ef, iter: Iterator[a, r]): Unit \ { ef, Read(r) } = match iter {
-        case Iterator(done, _) if done() => ()
-        case Iterator(_, next)           => f(next()); forEach(f, iter)
-    }
+    pub def forEach(f: a -> Unit \ ef1, iter: Iterator[a, ef, r]): Unit \ { ef, ef1, Read(r) } =
+        let Iterator(_, iterF) = iter;
+        def loop() = match iterF() {
+            case Skip   => loop()
+            case Ans(a) => f(a); loop()
+            case Done   => ()
+        };
+        loop()
 
     ///
     /// Applies `f` to every element of `iter` along with that element's index.
     ///
     /// Consumes the entire iterator.
     ///
-    pub def forEachWithIndex(f: (Int32, a) -> Unit \ ef, iter: Iterator[a, r]): Unit \ { ef, Read(r) } = region r {
-        let ix = ref 0 @ r;
-        forEach(x -> {let i = deref ix; f(i, x); ix := i + 1}, iter)
-    }
+    pub def forEachWithIndex(f: (Int32, a) -> Unit \ ef1, iter: Iterator[a, ef, r]): Unit \ { ef, ef1, Read(r) } =
+        let Iterator(_, iterF) = iter;
+        def loop(i) = match iterF() {
+            case Skip   => loop(i)
+            case Ans(a) => f(i, a); loop(i+1)
```

Comment:
Nifty.

Reply:
Thanks - no harm in using accumulators even if you have a region accessible.